### PR TITLE
Fix degenerate shape behavior for ord=+/-2

### LINKED ISSUE #59198: Output shape of `torch.linalg.norm` diverges from numpy equivalent in edge case.
## 🐛 Bug

Output shape of `torch.linalg.norm` diverges from numpy equivalent in edge case.

## To Reproduce

```python
import torch
import numpy as np

t = torch.rand((1, 1, 0))
a = t.numpy()

t_norm = torch.linalg.norm(t, ord=2, dim=(0, 1), keepdim=False)
a_norm = np.linalg.norm(a, ord=2, axis=(0, 1), keepdims=False)

assert t_norm.shape == a_norm.shape
```

- This also fails for `ord=-2` but works for all other orders
- This works fine for `keepdim=True`
- This also works fine for other degenerate input shapes 

## Environment

```
PyTorch version: 1.9.0a0+gitc9de3e3
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (crosstool-NG 1.24.0.133_b0863d8_dirty) 9.3.0
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.9

Python version: 3.6 (64-bit runtime)
Python platform: Linux-5.12.5-arch1-1-x86_64-with-arch
Is CUDA available: False
CUDA runtime version: 11.1.105
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080
Nvidia driver version: 465.31
cuDNN version: Probably one of the following:
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_adv_infer.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_adv_train.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_cnn_infer.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_cnn_train.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_ops_infer.so.8.1.1
/usr/local/cudnn-8.1.1-cuda-11/lib64/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] pytest-pytorch==0.0.0
[pip3] torch==1.9.0a0+gitc9de3e3
[conda] magma-cuda111             2.5.2                         1    pytorch
[conda] mkl                       2021.2.0           h726a3e6_389    conda-forge
[conda] mkl-include               2021.2.0           h726a3e6_389    conda-forge
[conda] numpy                     1.19.5           py36h2aa4a07_1    conda-forge
[conda] pytest-pytorch            0.2.0              pyh44b312d_0    conda-forge
[conda] torch                     1.9.0a0+gitc9de3e3           dev_0    <develop>
```

## Additional context

This is tested in 

https://github.com/pytorch/pytorch/blob/6d45d7a6c331ddb856ac34a76bcd3613aa05185b/test/test_linalg.py#L1928

but the test never failed due to a bug in the underlying comparison mechanism. #59067 will disable the failing test case, but it should be reinstated as soon as this issue is fixed.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano @VitalyFedyunin

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #59198

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -2167,31 +2167,13 @@ static void check_str_ord_valid(const c10::string_view str_ord, optional<IntArra
...
-    // the second reduction dim to 1, to avoid calling min/max, which would throw
-    // an error.
-    if (self.sizes()[dim] != 1) {
-      auto new_sizes = self.sizes().vec();
-      new_sizes[dim] = 1;
-      self.resize_(new_sizes);
-    }
-    result = keepdim ? self : self.flatten(dim);
+  if (ord > 0) {
+    return std::get<0>(self.max(dim, keepdim));
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebra.cpp:2173]: If indices tensor is not needed then we can use `self.amax(dim, keepdim)` here.

File: test/test_linalg.py [hunk 2]
```diff
@@ -1903,6 +1903,7 @@ def run_test_case(input, ord, dim, keepdim):
                 input_numpy = input.cpu().numpy()
                 result_numpy = np.linalg.norm(input_numpy, ord, dim, keepdim)
                 result = torch.linalg.norm(input, ord, dim, keepdim)
+                self.assertEqual(result.shape, result_numpy.shape, msg=msg)
```
Comment (IvanYashchuk) [inline test/test_linalg.py:1906]: Interestingly I faced exactly the same problem with `torch.linalg.cond` in https://github.com/pytorch/pytorch/pull/58041.
Maybe `self.assertEqual(torch.tensor, numpy.ndarray)` should check the shape as well.

**Review by IvanYashchuk — APPROVED** [review]

File: test/test_linalg.py (hunk 2 above)
Comment (kurtamohler) [reply test/test_linalg.py:1906]: I believe #59067 will add a shape check, and that's how @pmeier found this issue

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (kurtamohler) [reply aten/src/ATen/native/LinearAlgebra.cpp:2173]: I changed these to `amax` and `amin` when I initially submitted this PR, but I got some autograd errors in CI. They went away when I changed it back. https://app.circleci.com/pipelines/github/pytorch/pytorch/338771/workflows/9cc23b63-5edc-4a79-a100-9521e45f21f0/jobs/14238558

Comment (kurtamohler) [reply aten/src/ATen/native/LinearAlgebra.cpp:2173]: Actually, I guess avoiding inplace changes to the result of `amax` or `amin` should fix that error. Let me try it

Comment (kurtamohler) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Yep, that seemed to work

File: test/test_linalg.py (hunk 2 above)
Comment (pmeier) [reply test/test_linalg.py:1906]: > I believe #59067 will add a shape check

Kinda. It was already there, but never invoked. Now that #59067 has landed, you can remove this check and update the PR.

**Review by pmeier — COMMENTED** [review]
Since this PR fixes #59198, please also remove

https://github.com/pytorch/pytorch/blob/b0c9762e2d1dfcde549344628ad6be063378ef6a/test/test_linalg.py#L1973-L1977

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @kurtamohler!
